docs/tutorials/python/sagas/index.md

```diff
@@ -0,0 +1,586 @@
+The compensation, combined with the guarantee that the method will complete execution, makes this method a reliable, long-running transaction.
+
+In this guide, you'll build a Flask API that uses Temporal to manage the booking process for cars, hotels, and flights. 
+This approach ensures that even if one part of the booking fails, the system can gracefully handle the rollback of previous steps, maintaining data consistency.
+
+When you're finished, you'll be able to handle complex distributed transactions with ease and reliability using Temporal.
+
+## Prerequisites
+
+Before you begin, ensure you have the following:
+
+- Complete the [Hello World](/getting_started/python/hello_world_in_python/index.md) tutorial
+- Familiarity with asynchronous programming in Python
+- Basic understanding of microservices and distributed systems
+- The [Flask framework with async extras](https://flask.palletsprojects.com/en/2.3.x/async-await/) installed as a dependency of your project, which you can do with `pip install flask[async]`.
+
+## Create the booking functions
+
+You will start by creating the Activities for the booking process.
+These Activities form the core tasks your Workflow will perform, including interacting with external services and handling potential failures.
+Specifically, you'll create the booking Activities for cars, hotels, and flights.
+These Activities are used to interact with external services, but for this tutorial, you will not be making any actual service calls.
+Instead, you will stub these out and simulate failures by raising exceptions if a service is unavailable.
+First, create a new file named `activities.py`.
+This file will contain the definitions of the Activities needed for the booking process.
+
+Import the necessary modules:
+
+```python
+import asyncio
+from temporalio import activity
+from shared import BookVacationInput
+```
+
+The `asyncio` library is used for asynchronous operations.
+The `activity` module from the `temporalio` library provides decorators and functions for defining Activities.
+The `BookVacationInput` data class will be used to pass input data to the Activities.
+
+
+Next, define the `book_car`, `book_hotel`, and `book_flight` Activities.
+For brevity, these Activities will print a message indicating that they were invoked; however, in a real-world scenario, they would interact with external services to book the vacation.
+
+The function will return a success message if no errors occur.
+
+```python
+@activity.defn
+async def book_car(input: BookVacationInput) -> str:
+    await asyncio.sleep(3)
+    if activity.info().attempt < input.attempts:
+        activity.heartbeat(
+            f"Invoking activity, attempt number {activity.info().attempt}"
+        )
+        await asyncio.sleep(3)
+        raise RuntimeError("Car service is down")
+
+    if "invalid" in input.book_car_id:
+        raise Exception("Invalid car booking, rolling back!")
```

Comment:
Should not raise `Exception` directly, but something more like `RuntimeError` or something